R/aes-calculated.r

```diff
@@ -24,6 +26,41 @@
 stat <- function(x) {
   x
 }
+#' Mapped aesthetics
+#'
+#' Usually you will want the mapping of aesthetics to be based on input data
+#' but sometimes, e.g. for colour and fill, you'd want one aesthetic to be a
+#' function of another. Using the `mapped()` function you can mark an aesthetic
+#' to only be evaluated after all other aesthetics have been calculated and
+#' mapped. This allows you to e.g. set fill to be a transparent verison of ¨
+#' colour.
+#'
+#' @param x An aesthetic expression using only variables available after
+#'   mapping.
+#'
+#' @note Using `mapped()` for an aesthetic will not create a guide for that
+#' aesthetic, as it essentially a function of one of the other guides.
+#'
+#' @seealso [stat()] for marking aesthetics for evaluation after statistical
+#' transformation
+#' @export
+#' @examples
+#' ggplot(mpg, aes(class, hwy)) +
+#'   geom_boxplot(aes(colour = class, fill = mapped(alpha(colour, 0.4))))
+#'
+mapped <- function(x) {
+  x
+}
+#' @export
+stage <- function(stat = NULL, geom = NULL, mapped = NULL) {
+  stat
+}
+stage_geom <- function(stat = NULL, geom = NULL, mapped = NULL) {
+  geom
+}
+stage_mapped <- function(stat = NULL, geom = NULL, mapped = NULL) {
+  mapped
```

Comment:
Is `NULL` the right default here? I would've thought you'd want `waiver()`, but maybe I'm forgetting where that's used. The related change to `check_nondata_cols()` feels a bit suspicious to me.

Reply:
`waiver()` is not really used for aesthetics anywhere in the code and would require similar changes to `check_nondata_cols()` as well as additional places. `NULL` is the right call